openhtf/util/multicast.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Multicast facilities for sending and receiving messages.
+
+This module includes both a MulticastListener that listens on a multicast
+socket and invokes a callback function for each message recieved, and a send()
+function that is used to send one-shot messages to a multicast socket.
+"""
+
+
+import logging
+import socket
+import struct
+import threading
+
+
+_LOG = logging.getLogger(__name__)
+
+DEFAULT_ADDRESS = '239.1.1.1'
+DEFAULT_PORT = 10000
+DEFAULT_TTL = 1
+MAX_MESSAGE_BYTES = 1024  # Maximum allowable message length in bytes.
+PING_STRING = 'OPENHTF_PING'
+PING_RESPONSE_KEY = 'OPENHTF_PING_RESPONSE'
+
+
+class MulticastListener(threading.Thread):
+  """Agent that listens (and responds) to short messages on a multicast socket.
+
+  Args:
+    callback: A callable to invoke upon receipt of a multicast message. Will be
+              called with one argument -- the text of the message received.
+              callback can optionally return a string response, which will be
+              transmitted back to the sender.
+    address: Multicast IP address component of the socket to listen on.
+    port: Multicast UDP port component of the socket to listen on.
+    ttl: TTL for multicast messages. 1 to keep traffic in-network.
+  """
+  LISTEN_TIMEOUT_S = 3  # Seconds to listen before retrying.
+
+  def __init__(self,
+               callback,
+               address=DEFAULT_ADDRESS,
+               port=DEFAULT_PORT,
+               ttl=DEFAULT_TTL):
+    super(MulticastListener, self).__init__()
+    self.address = address
+    self.port = port
+    self.ttl = ttl
+    self._callback = callback
+    self._live = False
+    self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    self._sock.setsockopt(socket.IPPROTO_IP,
+                          socket.IP_MULTICAST_TTL,
+                          self.ttl)
+
+  def Stop(self):
+    """Stop listening for messages."""
+    self._live = False
+    self.join()
+
+  def run(self):
+    """Listen for pings until stopped."""
+    self._live = True
+    self._sock.bind(('', self.port))
+    self._sock.settimeout(self.LISTEN_TIMEOUT_S)
+    self._sock.setsockopt(
```

Comment:
Can you comment what this does? It's not clear from the code without looking all of this up